Source/CarthageKit/Project.swift

```diff
@@ -866,54 +858,56 @@ private func repositoryURLForProject(project: ProjectIdentifier, preferHTTPS: Bo
-			let cloneProducer: () -> SignalProducer<(ProjectEvent, NSURL), CarthageError> = {
-				let cloneProducer = cloneRepository(remoteURL, repositoryURL)
 
-				return SignalProducer(value: (.Cloning(project), repositoryURL))
-					.concat(cloneProducer.then(.empty))
-			}
-
-			if let _ = try? fileManager.createDirectoryAtURL(repositoryURL, withIntermediateDirectories: false, attributes: nil) {
-				// If we created the directory, we're now responsible for
-				// cloning it.
-				return cloneProducer()
-			} else {
-				return isGitRepository(repositoryURL)
-					.promoteErrors(CarthageError.self)
-					.flatMap(.Concat) { isRepository -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
-						if isRepository {
-							let fetchProducer = fetchRepository(repositoryURL, remoteURL: remoteURL, refspec: "+refs/heads/*:refs/heads/*") /* lol syntax highlighting */
-
-							return SignalProducer(value: (.Fetching(project), repositoryURL))
-								.concat(fetchProducer.then(.empty))
-						} else {
-							// If the directory isn't a repository (that might
-							// happen if the process is quitted right after
-							// creating the directory), remove the directory then
-							// clone it.
-							_ = try? fileManager.removeItemAtURL(repositoryURL)
-							return cloneProducer()
+		return .Success(repositoryURLForProject(project, preferHTTPS: preferHTTPS))
+	}
+	.flatMap(.Merge) { remoteURL -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
+		return isGitRepository(repositoryURL)
+			.promoteErrors(CarthageError.self)
+			.flatMap(.Merge) { isRepository -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
+				if isRepository {
+					let fetchProducer: () -> SignalProducer<(ProjectEvent, NSURL), CarthageError> = {
+						return SignalProducer(value: (.Fetching(project), repositoryURL))
+							.concat(fetchRepository(repositoryURL, remoteURL: remoteURL, refspec: "+refs/heads/*:refs/heads/*").then(.empty))
+					}
+					// If we've already cloned the repo, check for the revision, possibly skipping an unnecessary fetch
+					if let commitish = commitish {
+						return commitExistsInRepository(repositoryURL, revision: commitish)
+							.promoteErrors(CarthageError.self)
+							.flatMap(.Merge) { exists -> SignalProducer<(ProjectEvent, NSURL), CarthageError> in
+								if exists {
+									return SignalProducer(value: (.SkippedFetching(project), repositoryURL))
+								} else {
+									return fetchProducer()
+								}
 						}
+					} else {
+						return fetchProducer()
 					}
-			}
+				} else {
+					// Either the directory didn't exist or it did but wasn't a git repository
+					// (Could happen if the process is killed during a previous directory creation)
+					// So we remove it, then clone
+					_ = try? fileManager.removeItemAtURL(repositoryURL)
+					return SignalProducer(value: (.Cloning(project), repositoryURL))
+						.concat(cloneRepository(remoteURL, repositoryURL).then(.empty))
+				}
 		}
+	}
```

Comment:
Instead of this:

```swift
	return SignalProducer.attempt { () -> Result<GitURL, CarthageError> in
		do {
			try fileManager.createDirectoryAtURL(CarthageDependencyRepositoriesURL, withIntermediateDirectories: true, attributes: nil)
		} catch let error as NSError {
			return .Failure(.WriteFailed(CarthageDependencyRepositoriesURL, error))
		}

		return .Success(repositoryURLForProject(project, preferHTTPS: preferHTTPS))
	}
	.flatMap(.Merge) { remoteURL -> SignalProducer<(ProjectEvent, NSURL), CarthageError> 
		...
	}
```

We would typically write this:

```swift
	return SignalProducer.attempt { () -> Result<GitURL, CarthageError> in
			do {
				try fileManager.createDirectoryAtURL(CarthageDependencyRepositoriesURL, withIntermediateDirectories: true, attributes: nil)
			} catch let error as NSError {
				return .Failure(.WriteFailed(CarthageDependencyRepositoriesURL, error))
			}

			return .Success(repositoryURLForProject(project, preferHTTPS: preferHTTPS))
		}
		.flatMap(.Merge) { remoteURL -> SignalProducer<(ProjectEvent, NSURL), CarthageError> 
			...
		}
```